backend/pkg/kafka/consumer.go

```diff
@@ -99,100 +114,84 @@ func (s *Service) FetchMessages(ctx context.Context, progress IListMessagesProgr
 		kgo.ConsumePartitions(partitionOffsets),
 	)
 
-	// 4. Consume messages and check if filter code accepts the given records
+	// 4. Start go routine that consumes messages from Kafka and produces these records on the jobs channel so that these
+	// can be decoded by our workers.
+	go s.consumeKafkaMessages(workerCtx, client, consumeRequest, jobs)
+
+	// 5. Receive decoded messages until our request is satisfied. Once that's the case we will cancel the context
+	// that propagate to all the launched go routines.
 	messageCount := 0
 	remainingPartitionRequests := len(consumeRequest.Partitions)
+	for msg := range resultsCh {
+		// todo: Since a 'kafka message' is likely transmitted in compressed batches this is not really accurate
+		progress.OnMessageConsumed(msg.MessageSize)
+
+		if msg.IsMessageOk {
+			messageCount++
+			progress.OnMessage(msg)
+		}
+
+		partitionReq := consumeRequest.Partitions[msg.PartitionID]
+		if msg.Offset >= partitionReq.EndOffset {
+			remainingPartitionRequests--
+		}
+
+		// Do we need more messages to satisfy the user request? Return if request is satisfied
+		isRequestSatisfied := messageCount == consumeRequest.MaxMessageCount || remainingPartitionRequests == 0
+		if isRequestSatisfied {
+			// This will stop all the launched go routines for this consume request
+			cancel()
```

Comment:
This may as well move up to be a `defer cancel()` underneath where the cancel is declared.